File: apollo_dump/acquire_hook.py
# acquire_hook.py
# Python driver for acquire_hook.js
#
# Flow:
#   1. Auto-detect FreeStyle.exe PID
#   2. Inject acquire_hook.js via Frida
#   3. JS installs deception hooks, CRC patch, then ReadFile hook
#   4. When SSKF detected → scan stack for filename → cache target / replace source
#
# Usage:
#   py acquire_hook.py                     # auto PID + interactive
#   py acquire_hook.py <src> <dst>         # auto PID + set rule + wait
#   py acquire_hook.py --pid <PID>         # manual PID

import frida
import sys
import os
import time
import subprocess

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
JS_FILE = os.path.join(SCRIPT_DIR, 'acquire_hook.js')
LOG_FILE = os.path.join(SCRIPT_DIR, 'acquire_hook_log.txt')
_log_fh = None


def log_init():
    global _log_fh
    _log_fh = open(LOG_FILE, 'w', encoding='utf-8')
    _log_fh.write(f'=== acquire_hook log {time.strftime("%Y-%m-%d %H:%M:%S")} ===\n')
    _log_fh.flush()


def log_write(text):
    if _log_fh:
        _log_fh.write(text + '\n')
        _log_fh.flush()


def out(text):
    print(text)
    log_write(text)


def find_pid():
    try:
        result = subprocess.run(
            ['tasklist', '/fi', 'IMAGENAME eq FreeStyle.exe', '/fo', 'csv', '/nh'],
            capture_output=True, text=True, timeout=10
        )
        for line in result.stdout.strip().split('\n'):
            if 'FreeStyle.exe' in line:
                pid_str = line.split(',')[1].strip('"')
                return int(pid_str)
    except Exception as e:
        print(f'[!] tasklist failed: {e}')
    return None


def on_message(msg, data):
    if msg['type'] == 'send':
        payload = msg['payload']
        t = payload.get('t', '')

        if t == 'INIT':
            out(f'[*] {payload["msg"]}')
        elif t == 'SECTION':
            out(f'[*] .text: {payload["start"]} — {payload["end"]} ({payload["size"] / 1024 / 1024:.1f} MB)')
        elif t == 'DECEPTION':
            out(f'[STEP 0] {payload["msg"]}')
        elif t == 'DECEPTION_ERR':
            out(f'[!] DECEPTION ERROR: {payload["msg"]}')
        elif t == 'CRC':
            out(f'[STEP 0.5] {payload["msg"]}')
        elif t == 'CRC_ERR':
            out(f'[!] CRC ERROR: {payload["msg"]}')
        elif t == 'CRC_WARN':
            out(f'[*] CRC WARNING: {payload["msg"]}')
        elif t == 'READFILE':
            out(f'[STEP 1] {payload["msg"]}')
        elif t == 'READFILE_ERR':
            out(f'[!] READFILE ERROR: {payload["msg"]}')
        elif t == 'READY':
            out(f'\n{"="*60}')
            out(f'[READY] {payload["msg"]}')
            out(f'{"="*60}\n')
        elif t == 'SSKF':
            out(f'[SSKF #{payload["n"]}] {payload["size"]}B fname="{payload["fname"]}" (ESP+{payload["off"]:#x} {payload["type"]})')
        elif t == 'CACHE':
            out(f'[CACHE] {payload["item"]} → {payload["size"]}B cached from "{payload["fname"]}"')
        elif t == 'CACHE_ERR':
            out(f'[!] CACHE ERROR: {payload["err"]}')
        elif t == 'REPLACE':
            out(f'[REPLACE #{payload["count"]}] {payload["old"]} → {payload["new"]} ({payload["size"]}B)')
        elif t == 'REPLACE_SKIP':
            out(f'[SKIP] {payload["reason"]}: target={payload["targetSize"]}B buffer={payload["bufferSize"]}B')
        elif t == 'REPLACE_ERR':
            out(f'[!] REPLACE ERROR: {payload["err"]}')
        elif t == 'WAIT_TARGET':
            out(f'[*] {payload["msg"]}')
        elif t == 'CONFIG':
            out(f'[config] source="{payload["source"]}" target="{payload["target"]}"')
        elif t == 'STATUS':
            out(f'\n--- Status ---')
            out(f'  Source:     "{payload["source"]}"')
            out(f'  Target:     "{payload["target"]}"')
            out(f'  Replaces:   {payload["replaces"]}')
            out(f'  SSKF det:   {payload["sskf_detects"]}')
            out(f'  Cached:     {payload["cached"]}')
            out(f'  Deception:  active={payload.get("deception_active", False)}, VQ lies={payload.get("vq_lies", 0)}, VP fakes={payload.get("vp_fakes", 0)}')
            out(f'--------------\n')
        elif t == 'RESET':
            out(f'[reset] {payload["msg"]}')
        elif t == 'FAIL':
            out(f'\n[FAIL] {payload["msg"]}')
        else:
            out(f'[msg] {t}: {payload}')

    elif msg['type'] == 'error':
        out(f'[JS ERROR] {msg.get("description", msg)}')
        if 'stack' in msg:
            out(f'  Stack: {msg["stack"]}')


def interactive_loop(script):
    print('Commands:')
    print('  src <itemcode>       — Set source item code (e.g. src i50123971)')
    print('  dst <itemcode>       — Set target item code (e.g. dst i50125711)')
    print('  rule <src> <dst>     — Set both at once')
    print('  s / status           — Show status')
    print('  r / reset            — Reset counters')
    print('  q / quit             — Exit')
    print()

    while True:
        try:
            cmd = input('> ').strip()
        except (EOFError, KeyboardInterrupt):
            print('\n[exit]')
            break

        if not cmd:
            continue

        parts = cmd.split()
        op = parts[0].lower()

        if op in ('q', 'quit', 'exit'):
            break
        elif op == 'src' and len(parts) >= 2:
            script.exports.set_source(parts[1])
        elif op == 'dst' and len(parts) >= 2:
            script.exports.set_target(parts[1])
        elif op == 'rule' and len(parts) >= 3:
            script.exports.set_rule(parts[1], parts[2])
        elif op in ('s', 'status'):
            script.exports.status()
        elif op in ('r', 'reset'):
            script.exports.reset_counters()
        else:
            print(f'[?] Unknown command: {cmd}')
            print('    src <code> | dst <code> | rule <src> <dst> | s[tatus] | r[eset] | q[uit]')


def main():
    pid = None
    src_item = None
    dst_item = None

    args = sys.argv[1:]
    i = 0
    while i < len(args):
        if args[i] == '--pid' and i + 1 < len(args):
            pid = int(args[i + 1])
            i += 2
        elif not src_item and not args[i].startswith('--'):
            src_item = args[i]
            i += 1
        elif not dst_item and not args[i].startswith('--'):
            dst_item = args[i]
            i += 1
        elif args[i].startswith('--'):
            print(f'[!] Unknown flag: {args[i]}')
            i += 1
        else:
            i += 1

    if pid is None:
        print('[*] Finding FreeStyle.exe PID...')
        pid = find_pid()
        if pid is None:
            print('[!] FreeStyle.exe not found. Start the game first.')
            print('    Or specify --pid <PID>')
            sys.exit(1)
    print(f'[+] PID: {pid}')

    log_init()
    log_write(f'PID: {pid}')

    print(f'[*] Attaching to PID {pid}...')
    try:
        session = frida.attach(pid)
    except frida.ProcessNotFoundError:
        print(f'[!] Process {pid} not found')
        sys.exit(1)
    except Exception as e:
        print(f'[!] Attach failed: {e}')
        sys.exit(1)

    with open(JS_FILE, 'r', encoding='utf-8') as f:
        js_code = f.read()

    script = session.create_script(js_code)
    script.on('message', on_message)
    script.load()
    print(f'[+] Script loaded. Waiting for init...\n')

    time.sleep(2)

    if src_item and dst_item:
        time.sleep(1)
        try:
            script.exports.set_rule(src_item, dst_item)
        except Exception as e:
            print(f'[!] set_rule failed: {e}')

        print(f'[*] Rule set: {src_item} → {dst_item}')
        print(f'[*] Load TARGET item first (to cache), then load SOURCE item (to replace).')
        print(f'[*] Press Ctrl+C to exit.\n')
        try:
            while True:
                time.sleep(1)
        except KeyboardInterrupt:
            print('\n[exit]')
    else:
        interactive_loop(script)

    try:
        session.detach()
    except Exception:
        pass
    if _log_fh:
        _log_fh.close()
    print('[done]')


if __name__ == '__main__':
    main()

File: apollo_dump/_archived_20260605/acquire_hook.js
// acquire_hook.js
// Step 0:  Hook NtQueryVirtualMemory + NtProtectVirtualMemory (Apollo deception)
// Step 0.5: Patch ApolloCT CRC
// Step 1:  Hook ReadFile → detect SSKF → scan stack for filename → data replacement
//
// No AcquireSMD hook needed. Data replacement at ReadFile level:
// - TARGET item loads → cache its SSKF data
// - SOURCE item loads → overwrite buffer with TARGET data
//
// Usage: loaded by acquire_hook.py

'use strict';

var fsMod = Process.getModuleByName('FreeStyle.exe');
var base = fsMod.base;
var textSection = null;

fsMod.enumerateSections().forEach(function(s) {
    if (s.name === '.text') textSection = s;
});

if (!textSection) {
    send({t: 'FAIL', msg: '.text section not found in FreeStyle.exe'});
    throw new Error('.text section not found');
}
var textStart = textSection.address;
var textEnd = textStart.add(textSection.size);
send({t: 'SECTION', start: textStart.toString(), end: textEnd.toString(), size: textSection.size});

// ── Constants ────────────────────────────────────────────
var PAGE_EXECUTE_READ = 0x20;
var PAGE_EXECUTE_READWRITE = 0x40;

// ── Config ──────────────────────────────────────────────
var SOURCE_ITEM = '';
var TARGET_ITEM = '';
var REPLACE_COUNT = 0;

// ── Step 0/0.5 flags ─────────────────────────────────────
var VP_DISABLED = false;
var DECEPTION_ACTIVE = false;
var VQ_LIE_COUNT = 0;
var VP_FAKE_COUNT = 0;
var CRC_PATCHED = false;

// ── Helpers ──────────────────────────────────────────────
function inTextRange(addr) {
    return addr.compare(textStart) >= 0 && addr.compare(textEnd) < 0;
}

function addrFmt(a) {
    return a.toString() + '  (RVA: +' + a.sub(base).toString(16) + ')';
}

// ═══════════════════════════════════════════════════════════
// Step 0: Apollo deception hooks
// ═══════════════════════════════════════════════════════════

function installDeception() {
    send({t: 'DECEPTION', msg: 'Installing VirtualQuery/Protect deception hooks...'});

    var ntdll = Process.getModuleByName('ntdll.dll');

    // ── 0a: NtQueryVirtualMemory ─────────────────────────
    var NtQueryVirtualMemory = ntdll.getExportByName('NtQueryVirtualMemory');
    try {
        Interceptor.attach(NtQueryVirtualMemory, {
            onEnter: function(args) {
                this.memInfo    = args[3];
                this.infoLen    = args[4];
                this.infoClass  = args[2];
            },
            onLeave: function(retval) {
                if (retval.toInt32() !== 0) return;
                if (this.infoClass.toInt32() !== 0) return;
                if (this.memInfo.isNull()) return;
                try {
                    var baseAddr = this.memInfo.readPointer();
                    if (!inTextRange(baseAddr)) return;
                    this.memInfo.add(8).writeU32(PAGE_EXECUTE_READ);
                    this.memInfo.add(20).writeU32(PAGE_EXECUTE_READ);
                    VQ_LIE_COUNT++;
                } catch(e) {}
            }
        });
        send({t: 'DECEPTION', msg: '  NtQueryVirtualMemory hooked'});
    } catch(e) {
        send({t: 'DECEPTION_ERR', msg: 'NtQueryVirtualMemory: ' + e});
        return false;
    }

    // ── 0b: NtProtectVirtualMemory ────────────────────────
    var NtProtectVirtualMemory = ntdll.getExportByName('NtProtectVirtualMemory');
    try {
        Interceptor.attach(NtProtectVirtualMemory, {
            onEnter: function(args) {
                this.baseAddrPtr = args[1];
                this.oldProtectPtr = args[4];
                this.ourCall = VP_DISABLED;
            },
            onLeave: function(retval) {
                if (this.ourCall) return;
                try {
                    var addr = this.baseAddrPtr.readPointer();
                    if (!inTextRange(addr)) return;
                    if (!this.oldProtectPtr.isNull()) {
                        this.oldProtectPtr.writeU32(PAGE_EXECUTE_READ);
                    }
                    retval.replace(0);
                    VP_FAKE_COUNT++;
                } catch(e) {}
            }
        });
        send({t: 'DECEPTION', msg: '  NtProtectVirtualMemory hooked'});
    } catch(e) {
        send({t: 'DECEPTION_ERR', msg: 'NtProtectVirtualMemory: ' + e});
        return false;
    }

    DECEPTION_ACTIVE = true;
    send({t: 'DECEPTION', msg: 'Deception active.'});
    return true;
}

// ═══════════════════════════════════════════════════════════
// Step 0.5: Patch ApolloCT CRC
// ═══════════════════════════════════════════════════════════

function patchCRC() {
    send({t: 'CRC', msg: 'Patching ApolloCT CRC checks...'});

    var apolloCT;
    try {
        apolloCT = Process.getModuleByName('ApolloCT.dll');
    } catch(e) {
        send({t: 'CRC_ERR', msg: 'ApolloCT.dll not found'});
        return false;
    }

    var CRC_RVAS = [0x1A3C54, 0x1BE222];
    var patched = 0;

    for (var i = 0; i < CRC_RVAS.length; i++) {
        var addr = apolloCT.base.add(CRC_RVAS[i]);
        var page = addr.and(ptr(0xfffff000));
        try {
            Memory.protect(page, 0x1000, 'rwx');
            addr.writeU8(0x33);
            addr.add(1).writeU8(0xC0);
            addr.add(2).writeU8(0xC3);
            Memory.protect(page, 0x1000, 'rx');
            patched++;
            send({t: 'CRC', msg: '  Patched ApolloCT+0x' + CRC_RVAS[i].toString(16)});
        } catch(e) {
            send({t: 'CRC_ERR', msg: '  FAIL at 0x' + CRC_RVAS[i].toString(16) + ': ' + e});
        }
    }

    CRC_PATCHED = (patched === CRC_RVAS.length);
    send({t: 'CRC', msg: 'CRC patch: ' + patched + '/' + CRC_RVAS.length});
    return CRC_PATCHED;
}

// ═══════════════════════════════════════════════════════════
// Step 1: ReadFile hook → detect SSKF → scan stack → replace data
// ═══════════════════════════════════════════════════════════
// New approach: no AcquireSMD hook needed.
// - Scan stack for .smd filename when SSKF detected
// - Cache TARGET item's SSKF data when loaded
// - Replace SOURCE item's buffer with TARGET data

var READFILE_HOOKED = false;
var SSKF_CACHE = {};       // filename → { data: ArrayBuffer, size: int }
var SSKF_DETECT_COUNT = 0;

// Scan stack for .smd/.bml filename, return first match
function scanStackForFilename(esp) {
    for (var off = 0; off < 0x800; off += 4) {
        try {
            var ptrVal = esp.add(off).readPointer();
            // Deref: ptr → string
            var tries = [
                {fn: function() { return ptrVal.readAnsiString(128); }, type: 'deref'},
                {fn: function() { return ptrVal.add(4).readAnsiString(128); }, type: 'deref+4'},
                {fn: function() { return ptrVal.add(8).readAnsiString(128); }, type: 'deref+8'}
            ];
            for (var ti = 0; ti < tries.length; ti++) {
                try {
                    var s = tries[ti].fn();
                    if (s && (s.indexOf('.smd') >= 0 || s.indexOf('.bml') >= 0)) {
                        return { filename: s, off: off, type: tries[ti].type, addr: ptrVal.toString() };
                    }
                } catch(e) {}
            }
            // Double deref
            try {
                var ptr2 = ptrVal.readPointer();
                var s4 = ptr2.readAnsiString(128);
                if (s4 && (s4.indexOf('.smd') >= 0 || s4.indexOf('.bml') >= 0)) {
                    return { filename: s4, off: off, type: 'deref2', addr: ptrVal.toString() + '→' + ptr2.toString() };
                }
            } catch(e) {}
        } catch(e) {}
    }
    return null;
}

function installReadFileHook() {
    send({t: 'READFILE', msg: 'Hooking ReadFile for SSKF detection + data replacement...'});

    var kernel32 = Process.getModuleByName('kernel32.dll');
    var ReadFile = kernel32.getExportByName('ReadFile');

    try {
        Interceptor.attach(ReadFile, {
            onEnter: function(args) {
                this.hFile = args[0];
                this.lpBuffer = args[1];
                this.nNumberOfBytesToRead = args[2].toInt32();
                this.lpNumberOfBytesRead = args[3];
            },
            onLeave: function(retval) {
                if (retval.toInt32() === 0) return;
                if (this.nNumberOfBytesToRead < 4) return;
                if (this.lpBuffer.isNull()) return;

                // Check SSKF magic
                try {
                    var magic = this.lpBuffer.readByteArray(4);
                    if (!magic) return;
                    var m = new Uint8Array(magic);
                    if (!(m[0] === 0x53 && m[1] === 0x53 && m[2] === 0x4B && m[3] === 0x46)) return;
                } catch(e) { return; }

                SSKF_DETECT_COUNT++;

                // Get actual bytes read
                var bytesRead = this.nNumberOfBytesToRead;
                try {
                    var actual = this.lpNumberOfBytesRead.readU32();
                    if (actual > 0 && actual <= bytesRead) bytesRead = actual;
                } catch(e) {}

                // Scan stack for filename
                var esp = this.context.esp;
                var found = scanStackForFilename(esp);

                var fname = found ? found.filename : '(unknown)';
                send({t: 'SSKF', n: SSKF_DETECT_COUNT, size: bytesRead, fname: fname,
                    off: found ? found.off : -1, type: found ? found.type : ''});

                if (!found) return;

                // Cache or replace logic
                var isSource = SOURCE_ITEM && fname.indexOf(SOURCE_ITEM) >= 0;
                var isTarget = TARGET_ITEM && fname.indexOf(TARGET_ITEM) >= 0;

                if (isTarget && !SSKF_CACHE[TARGET_ITEM]) {
                    // Cache TARGET data
                    try {
                        var targetData = this.lpBuffer.readByteArray(bytesRead);
                        SSKF_CACHE[TARGET_ITEM] = { data: targetData, size: bytesRead };
                        send({t: 'CACHE', item: TARGET_ITEM, size: bytesRead, fname: fname});
                    } catch(e) {
                        send({t: 'CACHE_ERR', item: TARGET_ITEM, err: e.toString()});
                    }
                }

                if (isSource && SSKF_CACHE[TARGET_ITEM]) {
                    // Replace SOURCE buffer with TARGET data
                    var td = SSKF_CACHE[TARGET_ITEM];
                    if (td.size <= this.nNumberOfBytesToRead) {
                        try {
                            this.lpBuffer.writeByteArray(td.data);
                            // Update bytes read if different
                            try {
                                this.lpNumberOfBytesRead.writeU32(td.size);
                            } catch(e) {}
                            REPLACE_COUNT++;
                            send({t: 'REPLACE', count: REPLACE_COUNT, old: fname,
                                new: TARGET_ITEM, size: td.size});
                        } catch(e) {
                            send({t: 'REPLACE_ERR', err: e.toString()});
                        }
                    } else {
                        send({t: 'REPLACE_SKIP', reason: 'target larger than buffer',
                            targetSize: td.size, bufferSize: this.nNumberOfBytesToRead});
                    }
                }

                if (isSource && !SSKF_CACHE[TARGET_ITEM]) {
                    send({t: 'WAIT_TARGET', msg: 'Source detected but target not cached yet. Load target item first.'});
                }
            }
        });
        READFILE_HOOKED = true;
        send({t: 'READFILE', msg: 'ReadFile hooked. Waiting for SSKF reads...'});
        return true;
    } catch(e) {
        send({t: 'READFILE_ERR', msg: 'ReadFile hook failed: ' + e});
        return false;
    }
}

// ═══════════════════════════════════════════════════════════
// Init flow
// ═══════════════════════════════════════════════════════════

function init() {
    send({t: 'INIT', msg: 'Starting SSKF data replacement initialization...'});

    // Step 0: Apollo deception
    if (!installDeception()) {
        send({t: 'FAIL', msg: 'Deception hooks failed. Aborting.'});
        return;
    }

    // Step 0.5: CRC patch
    if (!patchCRC()) {
        send({t: 'CRC_WARN', msg: 'CRC patch failed, continuing...'});
    }

    // Step 1: ReadFile hook for SSKF detection + data replacement
    if (!installReadFileHook()) {
        send({t: 'FAIL', msg: 'ReadFile hook failed. Aborting.'});
        return;
    }

    send({
        t: 'READY',
        msg: 'Ready. Set rule: src <itemcode> dst <itemcode>. Load target item first to cache, then load source item to replace.'
    });
}

// ── RPC helpers ──────────────────────────────────────────
rpc.exports = {
    setSource: function(itemCode) {
        SOURCE_ITEM = itemCode;
        send({t: 'CONFIG', source: SOURCE_ITEM, target: TARGET_ITEM});
    },
    setTarget: function(itemCode) {
        TARGET_ITEM = itemCode;
        send({t: 'CONFIG', source: SOURCE_ITEM, target: TARGET_ITEM});
    },
    setRule: function(src, dst) {
        SOURCE_ITEM = src;
        TARGET_ITEM = dst;
        send({t: 'CONFIG', source: SOURCE_ITEM, target: TARGET_ITEM});
    },
    status: function() {
        var cached = [];
        var cacheKeys = Object.keys(SSKF_CACHE);
        for (var i = 0; i < cacheKeys.length; i++) {
            cached.push(cacheKeys[i] + '(' + SSKF_CACHE[cacheKeys[i]].size + 'B)');
        }
        send({
            t: 'STATUS',
            source: SOURCE_ITEM,
            target: TARGET_ITEM,
            replaces: REPLACE_COUNT,
            sskf_detects: SSKF_DETECT_COUNT,
            cached: cached.join(', ') || '(none)',
            deception_active: DECEPTION_ACTIVE,
            vq_lies: VQ_LIE_COUNT,
            vp_fakes: VP_FAKE_COUNT
        });
    },
    resetCounters: function() {
        REPLACE_COUNT = 0;
        send({t: 'RESET', msg: 'Counters reset'});
    }
};

// ── Start ────────────────────────────────────────────────
setTimeout(init, 500);
